pysd/py_backend/functions.py

```diff
@@ -434,46 +493,73 @@ def set_state(self, t, state):
+                if component_name[:7] == '_integ_':
+                    # we need to check the original expression to retrieve
+                    # the dimensions
+                    dims = literal_eval(re.findall("Subs: .+",
+                        getattr(self.components,
+                                component_name[7:]).__doc__)[0][6:])
+                else:
+                    dims = literal_eval(re.findall("Subs: .+",
+                        getattr(self.components,
+                                component_name).__doc__)[0][6:])
+            except Exception:
+                dims = None
+
+            if isinstance(value, np.ndarray):
+                # TODO: Remove when we have no backward compatible update
+                warnings.warn('using numpy.array for setting subscripted '
+                              + 'variables is deprecated, use a '
+                              + 'xarray.DataArray with the correct '
+                              + 'dimensions instead (https://pysd.readthedocs.io/en/master/basic_usage.html)',
+                              DeprecationWarning, stacklevel=2)
+                coords = {dim: self.components._subscript_dict[dim]
+                          for dim in dims}
+                value = xr.DataArray(value, coords, dims)
+
             # Try to update stateful component
             if hasattr(self.components, stateful_name):
                 try:
                     element = getattr(self.components, stateful_name)
+                    if dims:
+                        value = utils.rearrange(value, dims,
+                            self.components._subscript_dict)
                     element.update(value)
                 except AttributeError:
                     print("'%s' has no state elements, assignment failed")
                     raise
             else:
                 # Try to override component
                 try:
-                    setattr(self.components, component_name, self._constant_component(value))
+                    setattr(self.components, component_name,
+                            self._constant_component(value, dims))
                 except AttributeError:
                     print("'%s' has no component, assignment failed")
                     raise
 
-    def clear_caches(self):
-        """ Clears the Caches for all model elements """
-        for element_name in dir(self.components):
-            element = getattr(self.components, element_name)
-            if hasattr(element, 'cache_val'):
-                delattr(element, 'cache_val')
-
     def doc(self):
         """
-        Formats a table of documentation strings to help users remember variable names, and
-        understand how they are translated into python safe names.
+        Formats a table of documentation strings to help users remember
+        variable names, and understand how they are translated into
+        python safe names.
```

Comment:
can you add a description of the columns returned to the docstring? I think mostly so that when people who aren't familiar with subscripts get a result that has a subscript column, they know what it means...